### PR TITLE
🐛 Fix annotations.AddAnnotations (if annotations has been nil before)

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

Previously the `AddAnnotations` func only added the annotations in a local var in case the annotations map in the object has been nil before calling the func. 

As far as I could see this func is used in machine and machinepool controller when adding annotations to node resources. I suppose we didn't hit this bug because in those cases the annotation map is never nil.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: util/annotations/helpers.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func AddAnnotations(o metav1.Object, desired map[string]string) bool {
 	annotations := o.GetAnnotations()
 	if annotations == nil {
 		annotations = make(map[string]string)
+		o.SetAnnotations(annotations)
```
Comment (vincepri) [inline util/annotations/helpers.go:58]: THis should probably be set on line 65 only if the annotations have changed. We probably want to avoid turning a nil map into an instance if there are no changes?

Comment (vincepri) [reply util/annotations/helpers.go:58]: Also let's add a test to that effect?

Comment (CecileRobertMichon) [reply util/annotations/helpers.go:58]: I actually don't think we need this at all, maps are references in Go so modifying `annotations` will also modify the object annotation

File: util/annotations/helpers_test.go [hunk 2]
```diff
@@ -110,6 +110,23 @@ func TestAddAnnotations(t *testing.T) {
 			},
 			changed: true,
 		},
+		{
+			name: "should return true if annotations are changed and have been nil before",
```
Comment (CecileRobertMichon) [inline util/annotations/helpers_test.go:114]: is this test failing without your change?

File: util/annotations/helpers.go (hunk 1 above)
Comment (sbueringer) [reply util/annotations/helpers.go:58]: The problem is that we're just creating a new map and this doesn't affect the object we passed in. The added test case below fails without the code change here.

File: util/annotations/helpers_test.go (hunk 2 above)
Comment (sbueringer) [reply util/annotations/helpers_test.go:114]: yes

File: util/annotations/helpers.go (hunk 1 above)
Comment (sbueringer) [reply util/annotations/helpers.go:58]: @vincepri I could do it either here with the additional condition `len(desired) > 0` or below. Below I would also have to add some kind of check or do it for every annotation which we want to add. Both are fine for me.

I'll add a test so we don't change the nil to an empty map if we don't pass in annotations

Comment (sbueringer) [reply util/annotations/helpers.go:58]: Ah I'll just add the following at the top
```golang
if len(desired) == 0 {
  return false
}
```

Comment (vincepri) [reply util/annotations/helpers.go:58]: If we do set the annotations at the bottom, I think it should pass?

Comment (CecileRobertMichon) [reply util/annotations/helpers.go:61]: > The problem is that we're just creating a new map and this doesn't affect the object we passed in. The added test case below fails without the code change here.

oh I see now, you are right.

File: util/annotations/helpers_test.go [hunk 3]
```diff
@@ -69,6 +69,19 @@ func TestAddAnnotations(t *testing.T) {
 			},
 			changed: false,
 		},
+		{
+			name: "should do nothing if no annotations are provided and have been nil before",
```
Comment (CecileRobertMichon) [inline util/annotations/helpers_test.go:73]: 👍

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

File: util/annotations/helpers.go (hunk 1 above)
Comment (sbueringer) [reply util/annotations/helpers.go:61]: I added the guard clause above.
So the logic is like:
* if we don't need to do anything +> return false
* if the map is nil (and we know have to do something when len(desired)>0) => create and set annotation

But I can move it around if you prefer it another way.

**Review by vincepri — APPROVED** [review]
/approve
